### PR TITLE
[11.0] [mig] base_location: Migration to 11.0

### PR DESCRIPTION
**Pull request by etobella**
Migration to 11.0.

No major changes

### CODE REVIEW

File: base_location/demo/better_zip.xml
Comment (rgarnau) [inline base_location/demo/better_zip.xml]: change to "odoo" and remove "data"

File: base_location/models/better_zip.py
Comment (rgarnau) [inline base_location/models/better_zip.py]: deprecated

File: base_location/models/company.py
Comment (rgarnau) [inline base_location/models/company.py]: this goes after the fields

**Review by rgarnau — COMMENTED** [review]
LGTM :+1:  Just minor comments inline

File: base_location/README.rst
Comment (JordiBForgeFlow) [inline base_location/README.rst]: Adjust this section, to follow the same structure as in https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: base_location/README.rst
Comment (JordiBForgeFlow) [inline base_location/README.rst]: 11.0

File: base_location/README.rst
Comment (JordiBForgeFlow) [inline base_location/README.rst]: use https in all references in the readme

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Functional and technical review. Minor comments.

**Comment by pedrobaeza** [comment]
Odoo now contains a city_id field. Maybe we should adapt this module to the new core data model.

File: base_location/models/company.py
Comment (JordiBForgeFlow) [inline base_location/models/company.py]: res.company does not have a field 'city_id'

File: base_location/views/partner_view.xml
Comment (JordiBForgeFlow) [inline base_location/views/partner_view.xml]: The field should not be editable for contact persons, as it happens will other fields in this circumstance.
image


Just add `attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"`

**Review by JordiBForgeFlow — CHANGES_REQUESTED** [review]
Minor changes after adding city_id

File: base_location/models/better_zip.py
Comment (pedrobaeza) [inline base_location/models/better_zip.py]: command at the end for reducing diff on later changes

File: base_location/demo/better_zip.xml
Comment (pedrobaeza) [inline base_location/demo/better_zip.xml]: Isn't demo data in base for completing city_id field?

File: base_location/models/better_zip.py
Comment (pedrobaeza) [inline base_location/models/better_zip.py]: `readonly=True` for avoiding undesired writes on res.country

File: base_location/views/better_zip_view.xml
Comment (pedrobaeza) [inline base_location/views/better_zip_view.xml]: Don't split this attribute through several lines or the visualization inside UI will be ugly

File: base_location/views/better_zip_view.xml
Comment (pedrobaeza) [inline base_location/views/better_zip_view.xml]: If you hide this field, then hide also state field that is also auto-assigned from city

File: base_location/views/better_zip_view.xml
Comment (pedrobaeza) [inline base_location/views/better_zip_view.xml]: This should be dynamically added on country_id onchange, as this way, no state is shown if no country is selected.

File: base_location/views/better_zip_view.xml
Comment (pedrobaeza) [inline base_location/views/better_zip_view.xml]: This should be dynamically added on country_id onchange, as this way, no city is shown if no country is selected.

File: base_location/views/better_zip_view.xml
Comment (pedrobaeza) [inline base_location/views/better_zip_view.xml]: Don't change this only for stylistic reasons

File: base_location/views/better_zip_view.xml
Comment (pedrobaeza) [inline base_location/views/better_zip_view.xml]: Don't change this only for stylistic reasons

File: base_location/models/partner.py
Comment (pedrobaeza) [inline base_location/models/partner.py]: `self.city_id = self.zip_id.city_id`

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: base_location/demo/better_zip.xml
Comment (JordiBForgeFlow) [reply base_location/demo/better_zip.xml]: No, the module 'base_address_city' contains no demo data.

**Comment by pedrobaeza** [comment]
Please check CIs status

File: base_location/views/res_country_view.xml
Comment (HviorForgeFlow) [inline base_location/views/res_country_view.xml]: Already defined on module base_address_city.

File: base_location/views/better_zip_view.xml
Comment (HviorForgeFlow) [inline base_location/views/better_zip_view.xml]: Close tag

File: base_location/tests/test_completion.py
Comment (pedrobaeza) [inline base_location/tests/test_completion.py]: Why do you remove final commas for raising the diff lines and making in later changes to have 2 lines diff?

File: base_location/tests/test_completion.py
Comment (pedrobaeza) [inline base_location/tests/test_completion.py]: Please end all lines that are elements of a dictionary with the trailing comma as a good habit (you have added a lot with it).

File: base_location/models/partner.py
Comment (pedrobaeza) [inline base_location/models/partner.py]: s/better_zip_id/zip_id

File: base_location/models/company.py
Comment (pedrobaeza) [inline base_location/models/company.py]: This shouldn't be needed as company_id is inherited by delegation of res.partner. And please, don't call it `better_zip_id`.

File: base_location/models/company.py
Comment (pedrobaeza) [reply base_location/models/company.py]: You have to rename it to `self.zip_id.city_id`

File: base_location/models/partner.py
Comment (pedrobaeza) [inline base_location/models/partner.py]: Call it `ZIP location`

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Review by rgarnau — APPROVED** [review]

File: base_location/models/better_zip.py
Comment (yajo) [inline base_location/models/better_zip.py]: Shouldn't you be using a `res.lang`-specific template if available? Such as `%(city)s, %(state)s (%(country)s)` for example

File: base_location/models/better_zip.py
Comment (yajo) [inline base_location/models/better_zip.py]: Shouldn't this return `[]`?

File: base_location/models/better_zip.py
Comment (yajo) [inline base_location/models/better_zip.py]: Shouldn't this return `[]`?

File: base_location/models/partner.py
Comment (yajo) [inline base_location/models/partner.py]: Call it `_check_zip`

File: base_location/models/company.py
Comment (yajo) [inline base_location/models/company.py]: Add `oldname="better_zip_id"`

**Review by yajo — CHANGES_REQUESTED** [review]

**Review by JordiBForgeFlow — APPROVED** [review]

**Comment by JordiBForgeFlow** [comment]
@Yajo @pedrobaeza Can you do a last review? Seems that this one is ready to merge.

File: base_location/README.rst
Comment (pedrobaeza) [inline base_location/README.rst]: This should be in Configuration, and in the Usage section, explain that going to a partner, and filling the field "Autocomplete...", they will get the rest of the fields...

File: base_location/models/company.py
Comment (pedrobaeza) [inline base_location/models/company.py]: This should be a simply related field to `partner_id.zip_id`. Don't complicate things...

File: base_location/models/company.py
Comment (pedrobaeza) [inline base_location/models/company.py]: Why adding this field here? It's out of this module's scope.

File: base_location/models/company.py
Comment (pedrobaeza) [inline base_location/models/company.py]: `_onchange_zip_id`

File: base_location/models/company.py
Comment (pedrobaeza) [inline base_location/models/company.py]: Remove this method

File: base_location/models/company.py
Comment (pedrobaeza) [inline base_location/models/company.py]: Remove this method

File: base_location/models/company.py
Comment (pedrobaeza) [inline base_location/models/company.py]: Explain what this method does.

File: base_location/models/partner.py
Comment (pedrobaeza) [inline base_location/models/partner.py:29]: Is this really needed? Nobody searches zips only filtering the state: or you type the zip code or you type the city name

File: base_location/models/partner.py
Comment (pedrobaeza) [inline base_location/models/partner.py]: The same comment about its need. KISS principle.

File: base_location/models/partner.py
Comment (pedrobaeza) [inline base_location/models/partner.py]: Why putting this to False?

File: base_location/models/partner.py
Comment (pedrobaeza) [inline base_location/models/partner.py]: Put `for rec in self.filtered('zip_id'):` and you remove the need of putting it on the subsequent `if`

File: base_location/tests/test_base_location.py
Comment (pedrobaeza) [inline base_location/tests/test_base_location.py]: This is not Odoo flow: you have to create the company, and the partner is created from it.

File: base_location/models/better_zip.py
Comment (pedrobaeza) [inline base_location/models/better_zip.py]: Why splitting these methods here if they are used only one time?

File: base_location/models/better_zip.py
Comment (pedrobaeza) [inline base_location/models/better_zip.py]: No need of this initial `if`, as `self.state_id.country_id != self.country_id` is going to be handled correctly in case of country_id = False.

File: base_location/models/better_zip.py
Comment (pedrobaeza) [inline base_location/models/better_zip.py]: Handle here directly the domains (and look that the methods names are very fuzzy.

File: base_location/models/better_zip.py
Comment (pedrobaeza) [inline base_location/models/better_zip.py]: AFAIK, this is called automatically when you change the value of state_id.

**Review by pedrobaeza — DISMISSED** [review]

File: base_location/models/company.py
Comment (etobella) [reply base_location/models/company.py]: We are following the same configuration used in [odoo company](https://github.com/odoo/odoo/blob/11.0/odoo/addons/base/res/res_company.py#L59)

Comment (yajo) [reply base_location/models/company.py]: If I'm not wrong, `res.company` extends `res.partner`, so `zip_id` should be available out of the box, isn't it?

**Review by yajo — CHANGES_REQUESTED** [review]

File: base_location/models/company.py
Comment (yajo) [inline base_location/models/company.py]: Not sure what happens with GH, but I re-comment to get it displayed.

AFAIK `res.company` extends `res.partner`, so `zip_id` should be there out of the box, right?

Comment (pedrobaeza) [reply base_location/models/company.py]: Yeah, definitively I don't understand why the rest of the stuff would be needed having this easy approach of simply adding the field to the view, as in previous versions.

Comment (etobella) [reply base_location/models/company.py]: That is not true, as you can check on [odoo company definition](https://github.com/odoo/odoo/blob/11.0/odoo/addons/base/res/res_company.py#L11).

Comment (yajo) [reply base_location/models/company.py]: Well, indeed I was wrong: `res.company` does not extend `res.partner`. In any case, wouldn't it be enough to add the field as related instead of repeating the compute methods?

Comment (JordiBForgeFlow) [reply base_location/models/company.py]: @Yajo But odoo chose to duplicate the address fields into res.company. And we should respect that IMHO.

**Comment by etobella** [comment]
@pedrobaeza @Yajo is it necessary to change how it works? Or should we keep similar to the original address fields of `res.company`?

File: base_location/models/company.py
Comment (yajo) [reply base_location/models/company.py]: Yes, but [look at how it does it: basically reading and writing the field from the linked partner](https://github.com/odoo/odoo/blob/11.0/odoo/addons/base/res/res_company.py#L101-L130). That's basically a more complex way of using a related field, isn't it?

Comment (JordiBForgeFlow) [reply base_location/models/company.py]: @Yajo I propose that we stick to the way odoo works now. And maybe we can propose to odoo to change their approach upstream, and so when they fix in future versions we also fix.

**Comment by JordiBForgeFlow** [comment]
@Yajo @pedrobaeza Can we merge this one? I think that hitting 96.74% coverage is not too bad, right? :)

File: base_location/models/company.py
Comment (pedrobaeza) [reply base_location/models/company.py]: I don't agree. If there are no reasons for doing that way, we don't have to do the same as Odoo. Let's keep the code simple. The only possible reason I see if it's there's any issue when you first create a company. Please check that case. If not, KISS principle.

**Review by yajo — APPROVED** [review]
I don't really like overengineering things, and just doing that because Odoo does that doesn't seem a good decision IMHO. However, I won't block anymore on that matter.

File: base_location/models/company.py
Comment (JordiBForgeFlow) [reply base_location/models/company.py]: @pedrobaeza I have been testing and it is proving a mess to mix the two methods. When you make the fields related, the constrains that we added in res.partner, https://github.com/OCA/partner-contact/pull/488/files#diff-2f72fbcaf854c88b3e175783b321cd16R64, will fail. because when you change the city_id in the company, you are effectively changing it in the partner. The constrains on the partner are evaluated before the inverse methods update the other fields (city, etc..).

And we need constrains in the partner to ensure consistency.

So, as a conclusion, address fields are very related to each other. Either you make them all related to the partner in company, or you don't for all of them. But mixing methods produces inconsistencies.

Comment (pedrobaeza) [reply base_location/models/company.py]: Thanks for the explanation, Jordi. OK, that's a good reason for keeping things this way. That was the explanation I was looking for keeping things this way.

**Comment by pedrobaeza** [comment]
For me, 2 remaining things before merging:

- Put the explanation on the code about why is done this way instead with a related (for avoiding possible future questions). Jordi's explanation is very good.
- Squash useless commits from this migration.

File: base_location/models/company.py
Comment (yajo) [reply base_location/models/company.py]: Interesting. Maybe you could add a comment or known issue on that matter, to avoid future iterations on the same subject unless odoo changes its way...

File: base_location/models/company.py
Comment (JordiBForgeFlow) [inline base_location/models/company.py]: @pedrobaeza Comments here. I think that we can merge now.

File: base_location/models/better_zip.py
Comment (pedrobaeza) [reply base_location/models/better_zip.py]: This comment still remains

File: base_location/models/company.py
Comment (pedrobaeza) [reply base_location/models/company.py]: This comment still remains

File: base_location/models/company.py
Comment (pedrobaeza) [reply base_location/models/company.py]: This is not comment, but docstring, and can't be in the middle of the code. You should make here a real comment:
```
# In order to keep the same logic used in odoo, fields must be computed
# and inversed, not related. This way, we can ensure that it works
# ...
```

File: base_location/models/company.py
Comment (pedrobaeza) [reply base_location/models/company.py]: This comment still remains.

File: base_location/models/partner.py
Comment (pedrobaeza) [reply base_location/models/partner.py:29]: This comment still remains

File: base_location/models/partner.py
Comment (pedrobaeza) [reply base_location/models/partner.py]: This comment still remains

File: base_location/views/better_zip_view.xml
Comment (pedrobaeza) [reply base_location/views/better_zip_view.xml]: This comment still remains

File: base_location/models/company.py
Comment (etobella) [reply base_location/models/company.py]: City_id is not added on `base_address_city` module. So I had to add it in order to maintain its logic.

File: base_location/models/company.py
Comment (etobella) [reply base_location/models/company.py]: This method computes address fields from the partner. Used in `_compute_address` as shown in [company_code](https://github.com/odoo/odoo/blob/11.0/odoo/addons/base/res/res_company.py#L106)

File: base_location/models/partner.py
Comment (etobella) [reply base_location/models/partner.py:29]: It has been modified the domain filtering, but not the changes on fields (in order to avoid discrepancies)

File: base_location/models/better_zip.py
Comment (jesusVMayor) [inline base_location/models/better_zip.py]: What is the sense of this line? will never be executed

Comment (JordiBForgeFlow) [reply base_location/models/better_zip.py]: Why not? If the country is set to enforce the use of cities, it will.

Comment (jesusVMayor) [reply base_location/models/better_zip.py]: Because is "if NOT self.country_id.enforce_cities" and if the country is not set to enforce cities the field city_id is invisible and it can not be changed. So IMHO the if is unnecesary.

**Comment by yvaucher** [comment]
I don't want to be that guy but aren't we missing all the commits we had when the module was named `better_zip` ?

**Comment by yvaucher** [comment]
https://github.com/OCA/partner-contact/commits/10.0/better_zip

**Comment by etobella** [comment]
@yvaucher That is was related to the migration code because the module name was changed. I have changed it in order to recover them

**Comment by etobella** [comment]
@pedrobaeza I think that all comments have been modified now or an answer was provided.

**Comment by yvaucher** [comment]
@etobella nice work, thank you.

**Review by jesusVMayor — APPROVED** [review]
LGTM

**Comment by JordiBForgeFlow** [comment]
@pedrobaeza All comments have now been addressed. Can we merge?

**Pull request merged**
